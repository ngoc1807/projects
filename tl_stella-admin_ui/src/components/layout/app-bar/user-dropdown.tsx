import { FC, MouseEvent, useState } from "react";
import {
  Avatar,
  Menu,
  MenuItem,
  Button,
  useMediaQuery,
  ListItemIcon,
  useTheme,
} from "@mui/material";
import {
  KeyboardArrowDown as KeyboardArrowDownIcon,
  Logout as LogoutIcon,
} from "@mui/icons-material";
import { User } from "next-auth";

interface UserDropdownProps {
  user: User | undefined;
}

export const UserDropdown: FC<UserDropdownProps> = ({ user }) => {
  const theme = useTheme();
  const breakpointMediumAndUp = useMediaQuery(theme.breakpoints.up("md"));

  const [userMenuElement, setUserMenuElement] = useState<
    undefined | HTMLElement
  >();
  const openUserMenu = Boolean(userMenuElement);
  const handleUserMenuClick = (event: MouseEvent<HTMLElement>) => {
    setUserMenuElement(event.currentTarget);
  };
  const handleUserMenuClose = () => {
    setUserMenuElement(undefined);
  };

  const name = typeof user?.name === "string" ? user.name : undefined;
  const firstName =
    typeof user?.firstName === "string" ? user.firstName : undefined;
  const lastName =
    typeof user?.lastName === "string" ? user.lastName : undefined;

  const displayName =
    firstName && lastName ? `${firstName} ${lastName}` : name || "User";
  const initials =
    firstName && lastName
      ? `${firstName.charAt(0)}${lastName.charAt(0)}`
      : name?.charAt(0) ?? "U";

  return (
    <>
      <Button
        color="inherit"
        onClick={handleUserMenuClick}
        sx={{
          ml: 2,
          whiteSpace: "pre-wrap",
          textTransform: "none",
          letterSpacing: "normal",
          minWidth: 0,
        }}
      >
        <Avatar
          sx={{
            color: "white",
            backgroundColor: "primary.main",
            width: 30,
            height: 30,
            fontSize: 12,
            mr: 1,
          }}
        >
          {initials}
        </Avatar>
        {breakpointMediumAndUp ? `${displayName} ` : ` `}
        <KeyboardArrowDownIcon fontSize="small" />
      </Button>
      <Menu
        anchorEl={userMenuElement}
        open={openUserMenu}
        onClose={handleUserMenuClose}
        onClick={handleUserMenuClose}
        transformOrigin={{ horizontal: "right", vertical: "top" }}
        anchorOrigin={{ horizontal: "right", vertical: "bottom" }}
      >
        {/* <MenuItem>
          <ListItemIcon>
            <Settings fontSize="small" />
          </ListItemIcon>
          Settings
        </MenuItem> */}
        <li>
          <MenuItem component="a" href="/logout">
            <ListItemIcon>
              <LogoutIcon fontSize="small" />
            </ListItemIcon>
            Logout
          </MenuItem>
        </li>
      </Menu>
    </>
  );
};
