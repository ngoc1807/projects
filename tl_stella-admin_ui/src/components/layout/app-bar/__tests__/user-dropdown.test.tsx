import { cleanup, render } from "@testing-library/react";
import { UserDropdown } from "../user-dropdown";
import { User } from "next-auth";

const testUser = {
  id: "1",
  name: "test",
  firstName: "testing",
  lastName: "testy",
  email: "test@test.com",
  organizationID: "2",
  defaultTenantID: "3",
  isInternalUser: false,
};

describe("user-dropdown", () => {
  afterEach(() => {
    cleanup();
  });
  it("should render correctly when user is undefined", () => {
    const { container } = render(<UserDropdown user={undefined} />);
    expect(container.firstChild).toMatchSnapshot();
  });
  it("should render correctly when user is defined with only a name", () => {
    const user: User = { ...testUser, name: "Steve Rogers" };
    const { container } = render(<UserDropdown user={user} />);
    expect(container.firstChild).toMatchSnapshot();
  });
  it("should render correctly when user is defined with only a firstName", () => {
    const user: User = { ...testUser, firstName: "Steve" };
    const { container } = render(<UserDropdown user={user} />);
    expect(container.firstChild).toMatchSnapshot();
  });
  it("should render correctly when user is defined with only a lastName", () => {
    const user: User = { ...testUser, lastName: "Rogers" };
    const { container } = render(<UserDropdown user={user} />);
    expect(container.firstChild).toMatchSnapshot();
  });
  it("should render correctly when user is defined with a firstName and lastName but no name", () => {
    const user: User = { ...testUser, firstName: "Steve", lastName: "Rogers" };
    const { container } = render(<UserDropdown user={user} />);
    expect(container.firstChild).toMatchSnapshot();
  });
  it("should render correctly when user is defined with firstName, lastName, and name", () => {
    const user: User = {
      ...testUser,
      firstName: "Steve",
      lastName: "Rogers",
      name: "Steve Rogers",
    };
    const { container } = render(<UserDropdown user={user} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
