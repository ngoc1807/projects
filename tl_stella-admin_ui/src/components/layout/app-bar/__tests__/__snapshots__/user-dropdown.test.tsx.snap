// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user-dropdown should render correctly when user is defined with a firstName and lastName but no name 1`] = `
<button
  class="MuiButton-root MuiButton-text MuiButton-textInherit MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorInherit MuiButtonBase-root  css-jhzn3i-MuiButtonBase-root-MuiButton-root"
  tabindex="0"
  type="button"
>
  <div
    class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-v7bbez-MuiAvatar-root"
  >
    SR
  </div>
   
  <svg
    aria-hidden="true"
    class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
    data-testid="KeyboardArrowDownIcon"
    focusable="false"
    viewBox="0 0 24 24"
  >
    <path
      d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
    />
  </svg>
  <span
    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
  />
</button>
`;

exports[`user-dropdown should render correctly when user is defined with firstName, lastName, and name 1`] = `
<button
  class="MuiButton-root MuiButton-text MuiButton-textInherit MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorInherit MuiButtonBase-root  css-jhzn3i-MuiButtonBase-root-MuiButton-root"
  tabindex="0"
  type="button"
>
  <div
    class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-v7bbez-MuiAvatar-root"
  >
    SR
  </div>
   
  <svg
    aria-hidden="true"
    class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
    data-testid="KeyboardArrowDownIcon"
    focusable="false"
    viewBox="0 0 24 24"
  >
    <path
      d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
    />
  </svg>
  <span
    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
  />
</button>
`;

exports[`user-dropdown should render correctly when user is defined with only a firstName 1`] = `
<button
  class="MuiButton-root MuiButton-text MuiButton-textInherit MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorInherit MuiButtonBase-root  css-jhzn3i-MuiButtonBase-root-MuiButton-root"
  tabindex="0"
  type="button"
>
  <div
    class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-v7bbez-MuiAvatar-root"
  >
    St
  </div>
   
  <svg
    aria-hidden="true"
    class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
    data-testid="KeyboardArrowDownIcon"
    focusable="false"
    viewBox="0 0 24 24"
  >
    <path
      d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
    />
  </svg>
  <span
    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
  />
</button>
`;

exports[`user-dropdown should render correctly when user is defined with only a lastName 1`] = `
<button
  class="MuiButton-root MuiButton-text MuiButton-textInherit MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorInherit MuiButtonBase-root  css-jhzn3i-MuiButtonBase-root-MuiButton-root"
  tabindex="0"
  type="button"
>
  <div
    class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-v7bbez-MuiAvatar-root"
  >
    tR
  </div>
   
  <svg
    aria-hidden="true"
    class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
    data-testid="KeyboardArrowDownIcon"
    focusable="false"
    viewBox="0 0 24 24"
  >
    <path
      d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
    />
  </svg>
  <span
    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
  />
</button>
`;

exports[`user-dropdown should render correctly when user is defined with only a name 1`] = `
<button
  class="MuiButton-root MuiButton-text MuiButton-textInherit MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorInherit MuiButtonBase-root  css-jhzn3i-MuiButtonBase-root-MuiButton-root"
  tabindex="0"
  type="button"
>
  <div
    class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-v7bbez-MuiAvatar-root"
  >
    tt
  </div>
   
  <svg
    aria-hidden="true"
    class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
    data-testid="KeyboardArrowDownIcon"
    focusable="false"
    viewBox="0 0 24 24"
  >
    <path
      d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
    />
  </svg>
  <span
    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
  />
</button>
`;

exports[`user-dropdown should render correctly when user is undefined 1`] = `
<button
  class="MuiButton-root MuiButton-text MuiButton-textInherit MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorInherit MuiButtonBase-root  css-jhzn3i-MuiButtonBase-root-MuiButton-root"
  tabindex="0"
  type="button"
>
  <div
    class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-v7bbez-MuiAvatar-root"
  >
    U
  </div>
   
  <svg
    aria-hidden="true"
    class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
    data-testid="KeyboardArrowDownIcon"
    focusable="false"
    viewBox="0 0 24 24"
  >
    <path
      d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
    />
  </svg>
  <span
    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
  />
</button>
`;
